[Site Isolation] Add the most basic support for sharing a web process across sites

rniwa opened the pull request:
#### 0885d2d33b76ff339fc9d1987cc54b28bd838feb
<pre>
[Site Isolation] Add the most basic support for sharing a web process across sites
<a href="https://bugs.webkit.org/show_bug.cgi?id=296260">https://bugs.webkit.org/show_bug.cgi?id=296260</a>

Reviewed by Alex Christensen.

For performance reasons, it&apos;s desirable to be able to bundle websites in cross-site iframes into
a single Web process shared across sites even when site isolation is enabled. This PR introduces
a new runtime flag and implements the support for the most basic version of this concept. We create
at most one shared process for each BrowsingContextGroup.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml: Added a new runtime flag.
* Source/WebKit/UIProcess/BrowsingContextGroup.cpp:
(WebKit::BrowsingContextGroup::sharedBucketProcessForSite): Added. Returns the shared Web process
if it&apos;s applicable to this site and frame kind.
(WebKit::BrowsingContextGroup::ensureProcessForSite): Return the shared process&apos;s FrameProcess
if this is called with the shared process.
(WebKit::BrowsingContextGroup::processForSite):
(WebKit::BrowsingContextGroup::addFrameProcessAndInjectPageContextIf): Added the support for the
shared process.
(WebKit::BrowsingContextGroup::removeFrameProcess): Ditto.
* Source/WebKit/UIProcess/BrowsingContextGroup.h:
* Source/WebKit/UIProcess/FrameProcess.cpp:
(WebKit::FrameProcess::FrameProcess): Now takes const std::optional&lt;Site&gt;&amp; instead of const Site&amp;.
* Source/WebKit/UIProcess/FrameProcess.h:
(WebKit::FrameProcess::isSharedProcess const): Added.
(WebKit::FrameProcess::create): Reformatted.
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::receivedNavigationActionPolicyDecision): Use the shared process returned by
BrowsingContextGroup::sharedProcessForSite.
(WebKit::WebPageProxy::createNewPage): Fixed the code for the shared process.
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::didStartProvisionalLoadForMainFrame): Set m_site to an unexpected value
when there are multiple sites loaded in the process for one reason or another.
(WebKit::WebProcessProxy::didStartUsingProcessForSiteIsolation): Set m_site to
SiteState::SharedProcess if site is std::nullopt.
* Source/WebKit/UIProcess/WebProcessProxy.h:
(WebKit::WebProcessProxy::site):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm:
(TestWebKitAPI::enableFeature):
(TestWebKitAPI::enableSiteIsolation):
(TestWebKitAPI::siteIsolatedViewAndDelegate):
(TestWebKitAPI::siteIsolatedViewWithSharedProcess): Added.
(TestWebKitAPI::(SiteIsolation, SharedProcessMostBasic)):

Canonical link: <a href="https://commits.webkit.org/297753@main">https://commits.webkit.org/297753@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8c5a189e15b424e3fd4efeee72becf64f5df77ee

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/112728 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/32463 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/22941 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/118931 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/63230 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/114690 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/33115 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/41026 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/85800 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/36444 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/115675 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/26428 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/101417 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/66104 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/25728 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/19547 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/62690 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/105245 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/95822 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/19623 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/122152 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/111344 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/39806 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/29675 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/94662 "Passed tests") | 
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 services](https://ews-build.webkit.org/#/builders/28/builds/112224 "Build is in progress. Recent messages:") | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/40189 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/97644 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/94400 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/39521 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/17333 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/35909 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/18153 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/39692 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/45189 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/135574 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/39331 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/36425 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/42665 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/41070 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebKit/WebProcess/WebPage/WebFrame.cpp:
```diff
@@ -439,7 +439,7 @@ void WebFrame::createProvisionalFrame(ProvisionalFrameCreationParameters&& param
     RefPtr remoteFrame = coreRemoteFrame();
     if (!remoteFrame) {
-        ASSERT_NOT_REACHED();
+        ASSERT(coreLocalFrame() && parameters.isSlushBucketProcess);
```
This looks concerning.  This is an indication that the UI process isn't accounting for what frame is in what process correctly.

rniwa replied:
Well, the issue here is that having UI process be aware of which frame belongs to slush bucket would require new IPC roundtrips. I opted to account for that here instead so that `createProvisionalFrame` can be sent to the web process regardless of whether a frame belongs to the slush bucket or not.

rniwa replied:
Never mind. We can get rid of this code change.

achristensen07 commented on Source/WebKit/UIProcess/WebProcessProxy.cpp:
```diff
@@ -2232,8 +2232,15 @@ void WebProcessProxy::didStartProvisionalLoadForMainFrame(const URL& url)
 {
+    if (isSlushBucket == IsSlushBucket::Yes) {
+        if (!m_site)
+            m_site = site;
```
This doesn't seem quite right.  This needs a new state, like `Expected<IsSlushBucket, Site>` to mark the process as a slush bucket process instead of having a site or not yet having had a site assigned yet.

rniwa replied:
Sure, we can do that. I'll add a new enum `SiteState` which takes tri-states: `NotYetSpecified`, `MultipleSites`, `SlushBucket`.

achristensen07 commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -5052,6 +5052,12 @@ void WebPageProxy::receivedNavigationActionPolicyDecision(WebProcessProxy& proce
 
+    Ref browsingContextGroup = m_browsingContextGroup;
+    if (browsingContextGroup->shouldUseSlushBucketProcess(*this, Site { navigation->currentRequest().url() }, frame.isMainFrame() ? IsMainFrame::Yes : IsMainFrame::No)) {
+        continueWithProcessForNavigation(*browsingContextGroup->slushBucketProcess(), nullptr, "Found slush bucket process"_s);
```
This seems like it needs an ensureSlushBucketProcess that makes one if there isn't one.

rniwa replied:
On `BrowsingContextGroup`??

achristensen07 commented on Source/WebKit/UIProcess/BrowsingContextGroup.h:
```diff
@@ -73,6 +78,10 @@ class BrowsingContextGroup : public RefCountedAndCanMakeWeakPtr<BrowsingContextG
 
+    WeakPtr<WebProcessProxy> m_slushBucketProcess;
+    HashMap<WebCore::Site, WeakPtr<FrameProcess>> m_slushBucketProcessMap;
+    WeakHashSet<WebPageProxy> m_slushBucketPages;
```
This seems wrong to me.  Why is this needed?

rniwa replied:
Which part are you referring to?

achristensen07 commented on Source/WebKit/UIProcess/BrowsingContextGroup.cpp:
```diff
@@ -70,20 +93,28 @@ void BrowsingContextGroup::processDidTerminate(WebPageProxy& page, WebProcessPro
+    if (process.isSlushBucket()) {
+        if (m_slushBucketProcessMap.get(site) == &process)
+            return;
+        m_slushBucketProcessMap.set(site, process);
```
This looks like an unnecessary double hash lookup.

achristensen07 commented on Source/WebKit/UIProcess/BrowsingContextGroup.cpp:
```diff
@@ -42,9 +43,29 @@ BrowsingContextGroup::BrowsingContextGroup() = default;
 Ref<FrameProcess> BrowsingContextGroup::ensureProcessForSite(const Site& site, WebProcessProxy& process, const WebPreferences& preferences, InjectBrowsingContextIntoProcess injectBrowsingContextIntoProcess)
 {
     if (preferences.siteIsolationEnabled()) {
+        if (m_slushBucketSites.contains(site)) {
```
May as well null check m_slushBucketProcess here.

rniwa replied:
Sure, added.

achristensen07 commented on Source/WebKit/UIProcess/BrowsingContextGroup.cpp:
```diff
@@ -102,14 +118,35 @@ void BrowsingContextGroup::addFrameProcessAndInjectPageContextIf(FrameProcess& p
+    };
+
+    if (process.isSlushBucket()) {
+        ASSERT(m_slushBucketSites.size() == 1);
```
I don't see how this assertion would hold true for general browsing.

rniwa replied:
I think it should because when we're creating `FrameProcess` for the slush bucket, we've only added one site to the set but I'll change it to the loop over m_slushBucketSites anyway since it's equally easy to do.

achristensen07 commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -8729,7 +8736,10 @@ void WebPageProxy::createNewPage(IPC::Connection& connection, WindowFeatures&& w
             originatingFrameInfoData.frameID
         } });
-        configuration->setOpenedSite(openerFrame->frameProcess().site());
+        auto site = WebCore::Site(openerFrame->url());
```
nit: auto doesn't make anything better on this line.  May as well just specify the type, and you can reduce the number of keywords!

rniwa replied:
oh yeah, I'd use `WebCore::Site` on the LHS.

achristensen07 commented on Source/WebKit/UIProcess/BrowsingContextGroup.h:
```diff
@@ -25,11 +25,16 @@
 
 #pragma once
 
+#include "WebProcessProxy.h"
```
I think this can be a forward declaration instead of including a header from a header.

rniwa replied:
The issue is with `LockdownMode`, which is defined within `WebProcessProxy`.
I'll do a follow up to move it out of the class.

achristensen07 commented on Source/WebKit/UIProcess/BrowsingContextGroup.cpp:
```diff
@@ -102,14 +118,35 @@ void BrowsingContextGroup::addFrameProcessAndInjectPageContextIf(FrameProcess& p
-
+    if (process.isSlushBucket()) {
+        m_slushBucketProcess = nullptr;
+        m_slushBucketSites.clear();
```
I feel like we may want to have a mechanism for removing the sites one by one instead of having a growing collection that only clears when it is completely done.  This is fine for this PR, though.

rniwa replied:
Maybe? I don't think we can remove a site from slush bucket once it's added to it though unless all frames that had loaded the site has either been removed or navigated away.

rniwa committed:
[Site Isolation] Add the most basic support for sharing a web process across sites
https://bugs.webkit.org/show_bug.cgi?id=296260

Reviewed by Alex Christensen.

For performance reasons, it's desirable to be able to bundle websites in cross-site iframes into
a single Web process shared across sites even when site isolation is enabled. This PR introduces
a new runtime flag and implements the support for the most basic version of this concept. We create
at most one shared process for each BrowsingContextGroup.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml: Added a new runtime flag.
* Source/WebKit/UIProcess/BrowsingContextGroup.cpp:
(WebKit::BrowsingContextGroup::sharedBucketProcessForSite): Added. Returns the shared Web process
if it's applicable to this site and frame kind.
(WebKit::BrowsingContextGroup::ensureProcessForSite): Return the shared process's FrameProcess
if this is called with the shared process.
(WebKit::BrowsingContextGroup::processForSite):
(WebKit::BrowsingContextGroup::addFrameProcessAndInjectPageContextIf): Added the support for the
shared process.
(WebKit::BrowsingContextGroup::removeFrameProcess): Ditto.
* Source/WebKit/UIProcess/BrowsingContextGroup.h:
* Source/WebKit/UIProcess/FrameProcess.cpp:
(WebKit::FrameProcess::FrameProcess): Now takes const std::optional<Site>& instead of const Site&.
* Source/WebKit/UIProcess/FrameProcess.h:
(WebKit::FrameProcess::isSharedProcess const): Added.
(WebKit::FrameProcess::create): Reformatted.
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::receivedNavigationActionPolicyDecision): Use the shared process returned by
BrowsingContextGroup::sharedProcessForSite.
(WebKit::WebPageProxy::createNewPage): Fixed the code for the shared process.
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::didStartProvisionalLoadForMainFrame): Set m_site to an unexpected value
when there are multiple sites loaded in the process for one reason or another.
(WebKit::WebProcessProxy::didStartUsingProcessForSiteIsolation): Set m_site to
SiteState::SharedProcess if site is std::nullopt.
* Source/WebKit/UIProcess/WebProcessProxy.h:
(WebKit::WebProcessProxy::site):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm:
(TestWebKitAPI::enableFeature):
(TestWebKitAPI::enableSiteIsolation):
(TestWebKitAPI::siteIsolatedViewAndDelegate):
(TestWebKitAPI::siteIsolatedViewWithSharedProcess): Added.
(TestWebKitAPI::(SiteIsolation, SharedProcessMostBasic)):

Canonical link: https://commits.webkit.org/297753@main
